File: android/app/src/main/java/com/testlogon/android/feature/paper/PaperEngine.kt
package com.testlogon.android.feature.paper

import com.testlogon.android.data.exchange.OrderSide

/**
 * PURE, framework-free paper-trading engine -- the client-side simulation core. NO Android / coroutine /
 * Compose / network dependencies: every function below is a plain, deterministic transformation over
 * immutable value types so it can be unit-tested in isolation and called freely from a ViewModel.
 *
 * This engine NEVER touches the real order/matching stack -- it exists only to let a user practise with
 * simulated funds. All money / price / quantity are raw int64 engine units (Long), matching the rest of
 * the trading stack (price scalers are 1 today, so notional = price * qty). Position [qty] is SIGNED:
 * positive = long, negative = short.
 *
 * Accounting is AVERAGE-COST: extending a position folds the fill into a weighted-average entry; reducing
 * or closing realizes (fillPrice - avgEntry) * closedQty * dir (dir = +1 while long, -1 while short); a
 * fill larger than the open position cleanly flips it, re-seeding avgEntry at the fill price for the
 * residual. Cash moves on every fill (buy: cash -= price*qty ; sell: cash += price*qty), so a short sale
 * credits cash and a buy-to-cover debits it, exactly like the real venue's cash leg.
 */
object PaperEngine {

    /** Order type for a paper order. */
    enum class PaperOrderType { MARKET, LIMIT }

    /** Lifecycle of a paper order. MARKET orders are FILLED immediately; LIMIT orders start WORKING. */
    enum class PaperOrderStatus { WORKING, FILLED, CANCELLED }

    /**
     * A simulated order. [limitPrice] is required for LIMIT and ignored for MARKET. [qty] is always a
     * positive magnitude; [side] carries the direction. [id] is a caller-supplied stable identifier.
     */
    data class PaperOrder(
        val id: String,
        val symbolId: Int,
        val side: OrderSide,
        val type: PaperOrderType,
        val qty: Long,
        val limitPrice: Long? = null,
        val status: PaperOrderStatus = PaperOrderStatus.WORKING,
        val createdTsMs: Long = 0L,
    )

    /** A simulated fill (audit trail + history UI). [qty] is a positive magnitude; [side] the direction. */
    data class PaperFill(
        val orderId: String,
        val symbolId: Int,
        val side: OrderSide,
        val price: Long,
        val qty: Long,
        val tsMs: Long = 0L,
    )

    /** A held position. [qty] is SIGNED (negative = short). [avgEntry] is the average-cost basis (>=0). */
    data class PaperPosition(
        val qty: Long,
        val avgEntry: Long,
    )

    /**
     * The whole simulated account. [cash] is free simulated cash (can go negative under leverage/shorts,
     * matching a real cash leg). [positions] is keyed by symbolId; a symbol with a flat position is
     * removed. [realizedPnl] accumulates closed-trade PnL. [startingCash] is retained for the account
     * panel + reset baseline.
     */
    data class PaperAccount(
        val cash: Long,
        val positions: Map<Int, PaperPosition>,
        val orders: List<PaperOrder>,
        val fills: List<PaperFill>,
        val realizedPnl: Long,
        val startingCash: Long,
    )

    /** A fresh account seeded with [startingCash] free cash and nothing open. */
    fun newAccount(startingCash: Long): PaperAccount = PaperAccount(
        cash = startingCash,
        positions = emptyMap(),
        orders = emptyList(),
        fills = emptyList(),
        realizedPnl = 0L,
        startingCash = startingCash,
    )

    private fun dirOf(side: OrderSide): Int = if (side == OrderSide.BUY) 1 else -1

    /**
     * Place [order] against the current [marketPrice]. A MARKET order (and a LIMIT order already
     * marketable at placement) fills IMMEDIATELY: MARKET at [marketPrice], a marketable LIMIT at its
     * limit price (never worse than the limit). A non-marketable LIMIT is appended as a WORKING order to
     * be filled later by [onTick]. Orders with a non-positive qty, or a LIMIT with no limit price, are
     * ignored (returned unchanged) rather than corrupting the book.
     */
    fun placeOrder(acct: PaperAccount, order: PaperOrder, marketPrice: Long): PaperAccount {
        if (order.qty <= 0L) return acct
        return when (order.type) {
            PaperOrderType.MARKET ->
                fillOrder(acct, order.copy(status = PaperOrderStatus.FILLED), marketPrice)
            PaperOrderType.LIMIT -> {
                val limit = order.limitPrice ?: return acct
                if (isMarketable(order.side, limit, marketPrice)) {
                    // Marketable on arrival -- fill at the limit (price improvement kept by the venue-sim).
                    fillOrder(acct, order.copy(status = PaperOrderStatus.FILLED), limit)
                } else {
                    acct.copy(orders = acct.orders + order.copy(status = PaperOrderStatus.WORKING))
                }
            }
        }
    }

    /** A LIMIT is marketable when a BUY limit >= market, or a SELL limit <= market. */
    private fun isMarketable(side: OrderSide, limit: Long, marketPrice: Long): Boolean =
        if (side == OrderSide.BUY) marketPrice <= limit else marketPrice >= limit

    /**
     * Feed a new [marketPrice] for [symbolId] to the account, filling any WORKING LIMIT order on that
     * symbol that the price has reached: a BUY limit fills when marketPrice <= limit; a SELL limit fills
     * when marketPrice >= limit. Each triggered order fills at ITS OWN limit price (the sim assumes the
     * resting order is filled at its posted price). Orders are processed oldest-first.
     */
    fun onTick(acct: PaperAccount, symbolId: Int, marketPrice: Long): PaperAccount {
        var current = acct
        // Snapshot the working orders to fill so we iterate a stable list while `current` mutates.
        val toFill = acct.orders.filter {
            it.status == PaperOrderStatus.WORKING &&
                it.symbolId == symbolId &&
                it.type == PaperOrderType.LIMIT &&
                it.limitPrice != null &&
                triggers(it.side, it.limitPrice, marketPrice)
        }
        for (order in toFill) {
            val limit = order.limitPrice ?: continue
            current = fillOrder(current, order.copy(status = PaperOrderStatus.FILLED), limit)
        }
        return current
    }

    /** A resting BUY limit triggers at marketPrice <= limit; a resting SELL limit at marketPrice >= limit. */
    private fun triggers(side: OrderSide, limit: Long, marketPrice: Long): Boolean =
        if (side == OrderSide.BUY) marketPrice <= limit else marketPrice >= limit

    /**
     * Cancel the WORKING order with [orderId] (no-op if it's absent or already terminal). The order is
     * marked CANCELLED in place so it stays in history rather than vanishing.
     */
    fun cancelOrder(acct: PaperAccount, orderId: String): PaperAccount {
        val idx = acct.orders.indexOfFirst { it.id == orderId && it.status == PaperOrderStatus.WORKING }
        if (idx < 0) return acct
        val updated = acct.orders.toMutableList()
        updated[idx] = updated[idx].copy(status = PaperOrderStatus.CANCELLED)
        return acct.copy(orders = updated)
    }

    /**
     * Apply a fill of [order] at [fillPrice] to [acct]: average-cost position update + realized PnL +
     * cash leg + fill-history append + the order marked FILLED (replacing its WORKING entry when present,
     * else appended). The single choke point every fill path routes through.
     */
    private fun fillOrder(acct: PaperAccount, order: PaperOrder, fillPrice: Long): PaperAccount {
        val dir = dirOf(order.side)
        val signedQty = dir * order.qty
        val existing = acct.positions[order.symbolId] ?: PaperPosition(0L, 0L)

        var netQty = existing.qty
        var avgEntry = existing.avgEntry
        var realized = 0L

        if (netQty == 0L || sameSign(netQty, signedQty)) {
            // Opening or extending: weighted-average the entry over the (growing) absolute size.
            avgEntry = weightedEntry(netQty, avgEntry, order.qty, fillPrice)
            netQty += signedQty
        } else {
            // Opposite side: close against the open position (up to its size), realizing PnL.
            val closable = minOf(order.qty, Math.abs(netQty))
            val posDir = if (netQty > 0) 1 else -1
            realized = (fillPrice - avgEntry) * closable * posDir
            val remaining = Math.abs(netQty) - closable
            netQty = posDir.toLong() * remaining
            if (netQty == 0L) {
                avgEntry = 0L
                // Any residual beyond the close flips the position, re-seeded at the fill price.
                val flip = order.qty - closable
                if (flip > 0L) {
                    netQty = dir.toLong() * flip
                    avgEntry = fillPrice
                }
            }
        }

        // Cash leg: buy debits price*qty, sell credits price*qty (short sale credits, cover debits).
        val cashDelta = -signedQty * fillPrice
        val newPositions = acct.positions.toMutableMap()
        if (netQty == 0L) {
            newPositions.remove(order.symbolId)
        } else {
            newPositions[order.symbolId] = PaperPosition(qty = netQty, avgEntry = avgEntry)
        }

        val fill = PaperFill(
            orderId = order.id,
            symbolId = order.symbolId,
            side = order.side,
            price = fillPrice,
            qty = order.qty,
            tsMs = order.createdTsMs,
        )

        // Replace the working order in place (limit fill) or append (market / marketable-on-arrival).
        val orders = if (acct.orders.any { it.id == order.id }) {
            acct.orders.map { if (it.id == order.id) order else it }
        } else {
            acct.orders + order
        }

        return acct.copy(
            cash = acct.cash + cashDelta,
            positions = newPositions,
            orders = orders,
            fills = acct.fills + fill,
            realizedPnl = acct.realizedPnl + realized,
        )
    }

    private fun sameSign(a: Long, b: Long): Boolean = (a > 0 && b > 0) || (a < 0 && b < 0)

    /** Weighted-average entry when adding [addQty] @ [addPrice] onto |[heldQty]| @ [heldEntry]. */
    private fun weightedEntry(heldQty: Long, heldEntry: Long, addQty: Long, addPrice: Long): Long {
        val held = Math.abs(heldQty)
        val total = held + addQty
        if (total == 0L) return 0L
        return (held * heldEntry + addQty * addPrice) / total
    }

    /**
     * Mark-to-market value of a single [position] at [markPrice]: signed unrealized PnL =
     * (mark - avgEntry) * qty (qty already carries the sign, so a short profits as the mark falls).
     */
    fun positionMtm(position: PaperPosition, markPrice: Long): Long =
        (markPrice - position.avgEntry) * position.qty

    /**
     * Total unrealized PnL across all open positions, given a [marks] map of symbolId -> mark price.
     * A position with no available mark contributes 0 (unknown, not fabricated).
     */
    fun unrealized(acct: PaperAccount, marks: Map<Int, Long>): Long =
        acct.positions.entries.sumOf { (symbolId, pos) ->
            val mark = marks[symbolId] ?: return@sumOf 0L
            positionMtm(pos, mark)
        }

    /**
     * Account equity = free cash + mark value of every open position (avg-entry basis + its unrealized).
     * Equivalently cash + sum(qty * mark) over positions with a known mark; a position with no mark
     * contributes its cost basis only (mark treated as avgEntry -> 0 unrealized) so equity never jumps on
     * missing data.
     */
    fun equity(acct: PaperAccount, marks: Map<Int, Long>): Long =
        acct.cash + acct.positions.entries.sumOf { (symbolId, pos) ->
            val mark = marks[symbolId] ?: pos.avgEntry
            pos.qty * mark
        }
}

File: frontend/src/lib/paperEngine.ts
// Pure, framework-free client-side PAPER-TRADING engine. Everything stays in the
// exchange engine's int64 "tick" space (prices/qtys/money are integer ticks; the
// UI scales for display with the markets format helpers). NOTHING here touches
// the network or the real /me/orders — this is an isolated local simulation.
//
// Kept pure so it is unit-testable in isolation (see paperEngine.test.ts). Every
// mutating function returns a NEW account object (immutably); callers persist.

export type PaperSide = "buy" | "sell";
export type PaperOrderType = "market" | "limit";

/** A working (resting) or already-filled paper order. */
export interface PaperOrder {
  id: string;
  symbolId: number;
  side: PaperSide;
  type: PaperOrderType;
  /** Limit price in ticks (undefined for market orders). */
  price?: number;
  /** Order qty in ticks (always positive). */
  qty: number;
  status: "working" | "filled" | "cancelled";
  /** Creation time (ms epoch). */
  createdAt: number;
}

/** One executed paper fill. */
export interface PaperFill {
  id: string;
  orderId: string;
  symbolId: number;
  side: PaperSide;
  /** Fill price in ticks. */
  price: number;
  /** Fill qty in ticks (always positive). */
  qty: number;
  /** Realized PnL (ticks) attributed to this fill (0 when purely opening/adding). */
  realized: number;
  ts: number;
}

/** An open position under average-cost accounting. */
export interface PaperPosition {
  /** Signed qty in ticks — negative = short, positive = long, 0 = flat. */
  qty: number;
  /** Average entry price in ticks (0 when flat). */
  avgEntry: number;
}

export interface PaperAccount {
  /** Free quote cash in ticks. */
  cash: number;
  /** Open positions keyed by symbolId. */
  positions: Record<number, PaperPosition>;
  /** All orders (working + terminal), newest last. */
  orders: PaperOrder[];
  /** All fills, newest last. */
  fills: PaperFill[];
  /** Cumulative realized PnL in ticks. */
  realizedPnl: number;
  /** The cash the account was seeded with (for reset / return-% display). */
  startingCash: number;
}

/** Default seed: 100,000 quote units. Callers pass the value already in ticks. */
export const DEFAULT_STARTING_CASH = 100_000;

let idCounter = 0;
/** Monotonic-ish unique id; not persisted, only needs per-session uniqueness. */
function nextId(prefix: string): string {
  idCounter += 1;
  return `${prefix}_${Date.now().toString(36)}_${idCounter.toString(36)}`;
}

/** A fresh, empty paper account seeded with `startingCash` ticks. */
export function newAccount(startingCash: number = DEFAULT_STARTING_CASH): PaperAccount {
  const seed =
    Number.isFinite(startingCash) && startingCash > 0
      ? Math.floor(startingCash)
      : DEFAULT_STARTING_CASH;
  return {
    cash: seed,
    positions: {},
    orders: [],
    fills: [],
    realizedPnl: 0,
    startingCash: seed,
  };
}

/** Shallow-clone an account into a fresh mutable working copy. */
function cloneAccount(acct: PaperAccount): PaperAccount {
  const positions: Record<number, PaperPosition> = {};
  for (const [k, v] of Object.entries(acct.positions)) {
    positions[Number(k)] = { qty: v.qty, avgEntry: v.avgEntry };
  }
  return {
    cash: acct.cash,
    positions,
    orders: acct.orders.slice(),
    fills: acct.fills.slice(),
    realizedPnl: acct.realizedPnl,
    startingCash: acct.startingCash,
  };
}

/**
 * Apply one execution against the working account copy IN PLACE. Updates the
 * position via average-cost, realizes PnL when reducing/closing (with clean
 * flips), moves cash (buy: -price*qty, sell: +price*qty), and appends a fill.
 * Returns the realized PnL of this fill.
 */
function applyFill(
  acct: PaperAccount,
  order: PaperOrder,
  fillPrice: number,
  fillQty: number,
  ts: number,
): number {
  const side = order.side;
  const signed = side === "buy" ? fillQty : -fillQty;

  // Cash leg.
  acct.cash += side === "buy" ? -fillPrice * fillQty : fillPrice * fillQty;

  const pos = acct.positions[order.symbolId] ?? { qty: 0, avgEntry: 0 };
  const netQty = pos.qty;
  let realized = 0;

  if (netQty === 0 || Math.sign(signed) === Math.sign(netQty)) {
    // Opening from flat or adding in the same direction — weighted-average entry.
    const newQty = netQty + signed;
    const totalCost = pos.avgEntry * Math.abs(netQty) + fillPrice * Math.abs(signed);
    pos.avgEntry = Math.abs(newQty) === 0 ? 0 : totalCost / Math.abs(newQty);
    pos.qty = newQty;
  } else {
    // Opposite side — reduce/close, and maybe flip.
    const dir = netQty > 0 ? 1 : -1;
    const closedQty = Math.min(Math.abs(signed), Math.abs(netQty));
    realized = (fillPrice - pos.avgEntry) * closedQty * dir;
    acct.realizedPnl += realized;

    const residual = Math.abs(signed) - Math.abs(netQty);
    if (residual > 0) {
      // Closed the old leg entirely then flipped — open a fresh leg.
      pos.qty = Math.sign(signed) * residual;
      pos.avgEntry = fillPrice;
    } else {
      pos.qty = netQty + signed;
      if (pos.qty === 0) pos.avgEntry = 0;
    }
  }

  if (pos.qty === 0) {
    delete acct.positions[order.symbolId];
  } else {
    acct.positions[order.symbolId] = pos;
  }

  acct.fills.push({
    id: nextId("fill"),
    orderId: order.id,
    symbolId: order.symbolId,
    side,
    price: fillPrice,
    qty: fillQty,
    realized,
    ts,
  });

  return realized;
}

export interface PlaceOrderInput {
  symbolId: number;
  side: PaperSide;
  type: PaperOrderType;
  /** Required for limit orders; ignored for market. */
  price?: number;
  qty: number;
}

export interface PlaceOrderResult {
  account: PaperAccount;
  order: PaperOrder;
  /** The immediate fill, when the order executed on placement. */
  fill?: PaperFill;
}

/**
 * Place a paper order.
 * - MARKET: fills immediately at `marketPrice`.
 * - LIMIT: if it already crosses `marketPrice` it fills immediately at the
 *   limit price; otherwise it rests as a working order and fills later on tick.
 *   (buy crosses when marketPrice <= limit; sell crosses when marketPrice >= limit)
 */
export function placeOrder(
  acct: PaperAccount,
  input: PlaceOrderInput,
  marketPrice: number | undefined,
  ts: number = Date.now(),
): PlaceOrderResult {
  const next = cloneAccount(acct);
  const qty = Math.floor(input.qty);

  const order: PaperOrder = {
    id: nextId("ord"),
    symbolId: input.symbolId,
    side: input.side,
    type: input.type,
    price: input.type === "limit" ? input.price : undefined,
    qty,
    status: "working",
    createdAt: ts,
  };

  if (!(qty > 0)) {
    // Nothing to do — reject by returning a cancelled order, no state change.
    order.status = "cancelled";
    next.orders.push(order);
    return { account: next, order };
  }

  if (order.type === "market") {
    if (!(marketPrice != null && marketPrice > 0)) {
      // No market to fill against — cannot rest a market order, so cancel it.
      order.status = "cancelled";
      next.orders.push(order);
      return { account: next, order };
    }
    applyFill(next, order, marketPrice, qty, ts);
    order.status = "filled";
    next.orders.push(order);
    const fill = next.fills[next.fills.length - 1];
    return { account: next, order, fill };
  }

  // Limit order.
  const limit = order.price;
  if (!(limit != null && limit > 0)) {
    order.status = "cancelled";
    next.orders.push(order);
    return { account: next, order };
  }

  const crosses =
    marketPrice != null &&
    marketPrice > 0 &&
    ((order.side === "buy" && marketPrice <= limit) ||
      (order.side === "sell" && marketPrice >= limit));

  if (crosses) {
    applyFill(next, order, limit, qty, ts);
    order.status = "filled";
    next.orders.push(order);
    const fill = next.fills[next.fills.length - 1];
    return { account: next, order, fill };
  }

  // Rests as a working order.
  next.orders.push(order);
  return { account: next, order };
}

/**
 * Feed a new market price for `symbolId`; fills every working order the price now
 * satisfies (limit BUY fills when marketPrice <= limit; limit SELL fills when
 * marketPrice >= limit), each at ITS limit price. Returns a new account.
 */
export function onTick(
  acct: PaperAccount,
  symbolId: number,
  marketPrice: number | undefined,
  ts: number = Date.now(),
): PaperAccount {
  if (!(marketPrice != null && marketPrice > 0)) return acct;

  const working = acct.orders.filter(
    (o) =>
      o.status === "working" &&
      o.symbolId === symbolId &&
      o.type === "limit" &&
      o.price != null,
  );
  if (working.length === 0) return acct;

  const willFill = working.filter((o) =>
    o.side === "buy" ? marketPrice <= o.price! : marketPrice >= o.price!,
  );
  if (willFill.length === 0) return acct;

  const next = cloneAccount(acct);
  // Deterministic order: oldest working order first.
  const fillIds = new Set(willFill.map((o) => o.id));
  const ordered = next.orders
    .filter((o) => fillIds.has(o.id))
    .sort((a, b) => a.createdAt - b.createdAt);

  for (const o of ordered) {
    applyFill(next, o, o.price!, o.qty, ts);
    o.status = "filled";
  }
  return next;
}

/** Cancel a working order by id (no-op if not working). Returns a new account. */
export function cancelOrder(acct: PaperAccount, id: string): PaperAccount {
  const target = acct.orders.find((o) => o.id === id);
  if (!target || target.status !== "working") return acct;
  const next = cloneAccount(acct);
  const o = next.orders.find((x) => x.id === id)!;
  o.status = "cancelled";
  return next;
}

/** Reset to a fresh account, preserving the original starting cash by default. */
export function resetAccount(startingCash?: number): PaperAccount {
  return newAccount(startingCash ?? DEFAULT_STARTING_CASH);
}

/** Mark-to-market value of a single position at `mark` (ticks). */
export function positionMtm(pos: PaperPosition, mark: number | undefined): number {
  if (pos.qty === 0 || mark == null || !Number.isFinite(mark)) return 0;
  return pos.qty * mark;
}

/**
 * Unrealized PnL across all positions given a map of symbolId -> mark price.
 * A position with no mark contributes 0.
 */
export function unrealized(
  acct: PaperAccount,
  marks: Record<number, number | undefined>,
): number {
  let total = 0;
  for (const [k, pos] of Object.entries(acct.positions)) {
    const mark = marks[Number(k)];
    if (mark == null || !Number.isFinite(mark)) continue;
    total += (mark - pos.avgEntry) * pos.qty;
  }
  return total;
}

/** Account equity = cash + Σ position mark-to-market. */
export function equity(
  acct: PaperAccount,
  marks: Record<number, number | undefined>,
): number {
  let total = acct.cash;
  for (const [k, pos] of Object.entries(acct.positions)) {
    const mark = marks[Number(k)];
    if (mark == null || !Number.isFinite(mark)) continue;
    total += positionMtm(pos, mark);
  }
  return total;
}
